playbooks/roles/jenkins/tasks/plugins.yml

```diff
@@ -1,28 +1,121 @@
+
+# Below is copied, modified, from https://github.com/softasap/sa-box-jenkins/blob/master/tasks/plugins.yml
+# * * * *
+
+# Jenkins doesn't allow updates via CLI, though that is required before plugins
+# can be installed via CLI. See: https://gist.github.com/rowan-m/1026918
 - name: Create Jenkins updates folder.
   file:
-    path: /var/lib/jenkins/updates
+    path: "{{ jenkins_home }}/updates"
     owner: jenkins
     group: jenkins
     mode: 0755
     state: directory
 
 - name: Update Jenkins plugin data.
   shell: >
-    curl -L https://updates.jenkins-ci.org/update-center.json | sed '1d;$d' > /var/lib/jenkins/updates/default.json
-    creates=/var/lib/jenkins/updates/default.json
+    curl -L https://updates.jenkins-ci.org/update-center.json | sed '1d;$d' > "{{ jenkins_home }}/updates/default.json"
+    creates="{{ jenkins_home }}/updates/default.json"
 
 - name: Permissions for default.json updates info.
   file:
-    path: /var/lib/jenkins/updates/default.json
+    path: "{{ jenkins_home }}/updates/default.json"
     owner: jenkins
     group: jenkins
     mode: 0755
 
-# TODO: add -i option here so that it still works after setup
-- name: Install Jenkins plugins.
-  command: >
-    java -jar {{ jenkins_jar_location }} -s http://{{ jenkins_hostname }}:8080/ install-plugin {{ item }}
-    creates={{ jenkins_home }}/plugins/{{ item }}.jpi
-  with_items: jenkins_plugins
-  notify: restart jenkins
+- wait_for: port=8080
+
+- name: Get installed plugins
+  uri:
+    user: "{{ plugins_username }}"
+    password: "{{ plugins_password }}"
+    force_basic_auth: yes
+    url: "http://127.0.0.1:8080/pluginManager/api/json?tree=plugins[shortName]&{{ jenkins_crumb_token }}"
+    return_content: yes
+  register: jenkins_installed_plugins
+
+- name: Install plugins
+  uri:
+    user: "{{ plugins_username }}"
+    password: "{{ plugins_password }}"
+    force_basic_auth: yes
+    url: "http://127.0.0.1:8080/pluginManager/install?plugin.{{ item }}.default=on&{{ jenkins_crumb_token }}"
+    method: POST
+    status_code: [200, 302]
+  when: item not in jenkins_installed_plugins.json.plugins|map(attribute='shortName')|list
+  with_items: "{{ jenkins_plugins }}"
+
+- name: Wait for plugins to be installed
+  uri:
+    user: "{{ plugins_username }}"
+    password: "{{ plugins_password }}"
+    force_basic_auth: yes
+    url: "http://127.0.0.1:8080/updateCenter/installStatus?{{ jenkins_crumb_token }}"
+    return_content: yes
+  register: jenkins_plugin_status
+  until: "'Pending' not in jenkins_plugin_status.json.data.jobs|map(attribute='installStatus')"
+  retries: 120
+  delay: 10
+
+- name: Check if we need to restart Jenkins to activate plugins
+  uri:
+    user: "{{ plugins_username }}"
+    password: "{{ plugins_password }}"
+    force_basic_auth: yes
+    url: "http://127.0.0.1:8080/updateCenter/api/json\
+    ?tree=restartRequiredForCompletion&{{ jenkins_crumb_token }}"
+    return_content: yes
+  register: jenkins_restart_required
```

Comment:
Could we put this in an Ansible `notify` handler for restarting Jenkins rather than doing the `wait` stuff above? Or does this have to be done before the next task?

Reply:
Yeah, this needs to be done before the next tasks are run when booting from scratch. If you just run all the tags in one go, the handlers only get run at the end of everything.

Reply:
Makes sense!